File: model/tools/filt_step.cpp
// C++ replacement for filt_step.py. Exports the existing filt_cyc captures with
// their verified onsets; byte-identical step.bin and decay.bin, no Python/FPU.
#include "filt_capture.h"
#include <filesystem>
int main(){
 std::filesystem::create_directories("work/filt");
 FILE *s=fopen("work/filt/step.bin","wb"),*d=fopen("work/filt/decay.bin","wb");
 if(!s||!d)return 1;
 for(int i=0;i<24;i++){
  const auto&cf=cyc_config[i];auto c=cap("tests/filt_cyc/hw/fc_"+std::to_string(i/4));
  if(c.ns!=4||c.n<unsigned(cf.on+1650))return 2;
  std::vector<int32_t> y(c.n);
  for(unsigned n=0;n<c.n;n++){int32_t v=c.v[n*4+i%4];if(v&1)return 3;y[n]=-v/2;}
  int32_t hs[]={cf.F,cf.Q,cf.A,cf.on,(int32_t)c.n},hd[]={cf.F,cf.Q,cf.A,1500};
  if(fwrite(hs,4,5,s)!=5||fwrite(y.data(),4,y.size(),s)!=y.size()||
     fwrite(hd,4,4,d)!=4||fwrite(y.data()+cf.on+150,4,1500,d)!=1500)return 4;
 }
 bool ok=fclose(s)==0;ok=(fclose(d)==0)&&ok;
 puts("exported 24 streams to work/filt/{step,decay}.bin");return !ok;
}

File: model/tools/filt_step.py
#!/usr/bin/env python3
"""filt_step.py -- export the filt_cyc step responses for the filter searches.

tests/filt_cyc/hw/fc_<b>.{hdr,bin} stream k = filt_cyc case b*4+k (F, Q, A: rest -> 256 samples of A -> zeros).
Writes work/filt/step.bin: per stream int32 F, Q, A, on, N, y[N] with y = -MIXS/2 (1/8 sample units; MIXS is always
even with VOFF = 1) and `on` = the first sample that can see the input (filt_lp.ON: the first output change after
key-on -- 139 for batch 0, 138 for batches 1..4, 200 for batch 5).  Also writes work/filt/decay.bin (the older
DC-to-zero windows: per stream F, Q, A, N=1500, y[on+150 .. on+1650)) for filt_search6..8.
"""
import os, struct, sys
import numpy as np
sys.path.insert(0, os.path.dirname(__file__))
import cap
from filt_lp import CF, ON

def main():
    os.makedirs("work/filt", exist_ok=True)
    with open("work/filt/step.bin", "wb") as fs, open("work/filt/decay.bin", "wb") as fd:
        for i, (F, Q, A) in enumerate(CF):
            d = cap.load(f"tests/filt_cyc/hw/fc_{i // 4}")["data"][:, i % 4].astype(np.int64)
            assert not np.any(d & 1)
            y = (-d // 2).astype("<i4")
            fs.write(struct.pack("<5i", F, Q, A, ON[i], len(y)) + y.tobytes())
            w = y[ON[i] + 150:ON[i] + 1650]
            fd.write(struct.pack("<4i", F, Q, A, len(w)) + w.tobytes())
    print(f"work/filt/step.bin, work/filt/decay.bin: {len(CF)} streams")

if __name__ == "__main__":
    main()
